src/libsyntax/ast.rs

```diff
@@ -1254,6 +1254,23 @@ pub enum ExprKind {
     Err,
 }
 
+impl ExprKind {
+    /// Whether this expression can appear in a contst argument without being surrounded by braces.
+    ///
+    /// Only used in error recovery.
+    pub(crate) fn is_valid_const_on_its_own(&self) -> bool {
+        match self {
+            ExprKind::Tup(_) |
+            ExprKind::Lit(_) |
+            ExprKind::Type(..) |
```

Comment:
Not sure about this one as well; `id::< 1 : u8 >()` won't parse. I'd add some more test cases.